Navigation Block: add block gap support

ramonjd opened the pull request:
## Description
https://github.com/WordPress/gutenberg/pull/33991 has made block gap support available 🎉 

Here we're applying block spacing support to the Navigation Block. More specifically, to the inner flex container surrounding nav items and page list items.

The calculation is to (try to) maintain the original proportional differences between row and column gaps.

Resolves https://github.com/WordPress/gutenberg/issues/34525

## Testing

This might take a while, so grab a drink, get comfortable. Some lo-fi beats often help me, or heavy metal, depending on the weather.

Enable block gap/spacing support in your theme.json under "settings", e.g., 

```json
		"spacing": {
			"blockGap": true,
                         ...

```

Have some posts and pages ready for testing your nav! Also if there are any default `blockGap` spacing values in your `theme.json` ( e.g., `"spacing": { "blockGap": "24px" }`) remove them for now so we can check the CSS defaults.

Insert a Navigation Block, start with an empty menu and add post/page items. 

In the Dimensions Panel in the controls sidebar, adjust the block spacing value.

Repeat the above steps and add a Navigation block, selecting **Add all pages**.

<img width="307" alt="Screen Shot 2021-10-01 at 11 53 16 am" src="https://user-images.githubusercontent.com/6458278/135553492-79487478-e278-450e-8f81-3c9ea66760ff.png">

Check that the block spacing value is applied to the menu items in the editor and for the saved post on the frontend.

Remove all `blockGap` values. Delete them from the input rather than setting to `0`.

Add background colors to your Navigation Blocks.

Check that the block defaults match those defaults in the CSS.

Now, in your theme.json, add the following to the "styles" object:


```json
		"blocks": {
			"core/navigation": {
				"spacing": {
					"blockGap": "25px"
				}
			}
		}
```

Check that the default value of `--wp--style--block-gap` is `25px` and that it is applied to the Navigation container.

Enable responsive menu in the controls sidebar:

<img width="273" alt="Screen Shot 2021-10-01 at 11 50 44 am" src="https://user-images.githubusercontent.com/6458278/135553319-38b71126-9e91-4614-9648-7d6e790e9ece.png">
 
Ensure that the block spacing still applies to the menus in wider-width screens (not the overlay)

## Screenshots


![navigate](https://user-images.githubusercontent.com/6458278/136135636-380fb722-41d4-4150-9d29-ae510b7211a0.gif)

## Types of changes
Adds block spacing support to the navigation block

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ramonjd committed:
Initial commit. Here we're applying block gap support to the inner flex container surrounding nav items and page list items.
The calculation is to (try to) maintain the original proportional differences between row and column gaps.

ramonjd committed:
Setting the default to zero, as it was

jasmussen commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -230,7 +230,8 @@
 .wp-block-navigation .wp-block-page-list,
 .wp-block-navigation__container {
-	gap: 0.5em 2em;
+	row-gap: calc(var(--wp--style--block-gap, 2em) / 4);
```
A thought, since we are thinking about a block spacing control that in the future allows you to swap between uniform and axial gap, perhaps we should keep the single `gap` property, but just switch between the singular and shorthand syntax?

From [CSS Tricks](https://css-tricks.com/almanac/properties/g/gap/):

```
.container {
  gap: 1rem;
  /* Is equivalent to:
  *  row-gap: 1rem;
  *  column-gap: 1rem
  */
  
  gap: 10px 15%;
  /* Is equivalent to:
  *  row-gap: 10px;
  *  column-gap: 15%;
  */
}
```

<img width="1012" alt="Spacing" src="https://user-images.githubusercontent.com/1204802/135582805-000315b8-80b8-4b81-b4ff-36deea7a0578.png">

ramonjd replied:
> A thought, since we are thinking about a block spacing control that in the future allows you to swap between uniform and axial gap, perhaps we should keep the single gap property, but just switch between the singular and shorthand syntax?

Thanks for bringing the uniform/axial controls to our attention.

I went with longhand just to make things explicit for me at the time, while I was thinking of how to implement decent defaults. 

Though shorthand is just as good 👍 

I was approaching the task with a view of trying to maintain the current defaults:

```
// Defaults to gap: 0.5em 2em where --wp--style--block-gap isn't available.
.wp-block-navigation__container {
    gap: calc(var(--wp--style--block-gap, 2em) / 4) var(--wp--style--block-gap, 2em);
}
```

When axial eventually becomes a requirement, I see how this won't work however. 

Did you have a method of switching between the singular and shorthand syntax in mind?

My first thought is that we could move the logic to the backend (either by introducing new CSS vars or inline styles). It would make things easier to calculate conditionally.

ramonjd replied:
> Did you have a method of switching between the singular and shorthand syntax in mind?

Actually, sorry, just saw the ongoing discussions here:

https://github.com/WordPress/gutenberg/issues/34529

cc @stacimc @andrewserong who have been thinking deeply about this topic.

jasmussen replied:
Yep the idea is that you'll be able to switch between uniform or axial gap, with most (all?) blocks defaulting to uniform gap. 

Embracing the singular `gap` value might also be good for global styles overrides, since you could use a single value to support either. But the inner workings here are a bit unclear for me still, so I'm happy to defer.

jasmussen commented:
Very nice, thank you! Here's a gif of it working:
![gap](https://user-images.githubusercontent.com/1204802/135583413-d4d9087e-cd28-4cae-83c2-a62d899fc847.gif)

One of the original use cases was a space-betweene'd pattern with a site logo in the middle. Works great:

![site logo](https://user-images.githubusercontent.com/1204802/135583967-aff0bd22-b5df-4838-aa57-8f843ea0bfce.gif)

- Perhaps this originated from a separate PR, but let's call the control "Block spacing" instead of "Block gap".
- Let's definitely show the block spacing as a default control for Navigation.

In general for the new ToolsPanel, I think the value it has is that it an _curate_ the controls inside, which also means it should, with few exceptions, never be empty. The idea is, if there are 10 controls available (a la typography), we can show the 4 most valuable ones and ensure there's still access to the remaining ones. If there's only a single control available, we'll almost always want to just show it by default. 

What do you think?

andrewserong commented:
> Perhaps this originated from a separate PR, but let's call the control "Block spacing" instead of "Block gap".

Yes, the rename is included in the [PR](https://github.com/WordPress/gutenberg/pull/34630) for the Columns block gap.

ramonjd commented:
>  If there's only a single control available, we'll almost always want to just show it by default.

Definitely. Great point. 👍 

>  let's call the control "Block spacing" instead of "Block gap".

👍 Sounds good. 

Just like the start of a new year, it will take me a few days to reprogram my 🧠 to write it correctly 😆

ramonjd committed:
Using shorthand styles
Ensuring that spacing gap controls are shown by default in the otherwise empty Dimensions panel.

tellthemachines reviewed (commented):
This is working well! 

To clarify, with `Check that the default value of --wp--style--block-gap is 25px and that it is applied to the Social Links Block container.` did you mean is _isn't_ applied to Social Links? Given the setting above is specific to the Navigation block. Fwiw I tested and it doesn't apply to Social Links 😅 

I also tested with a really old theme that doesn't set a value for `--wp--style--block-gap` and the fallback value is working correctly ✅

ramonjd commented:
> Check that the default value of --wp--style--block-gap is 25px and that it is applied to the Social Links Block container. did you mean is isn't applied to Social Links? Given the setting above is specific to the Navigation block. Fwiw I tested and it doesn't apply to Social Links 😅

I put that in there just to make sure people were paying attention :trollface: 

Seriously, thanks for testing and spotting that copy paste error (from https://github.com/WordPress/gutenberg/pull/35236)

I've updated the test description. It should indeed refer to the Navigation block, not the Social Links block.

tellthemachines reviewed (approved):
Awesome, I think this is good to go then ✅

ramonjd commented:
Quick rebase before merging to be able to test social icons gap spacing within a navigation block.

![nav-gap](https://user-images.githubusercontent.com/6458278/137827697-78f6f040-33b1-443e-99cc-32f95d319a51.gif)

👍

jasmussen commented:
Thanks for working on this!

I'm seeing in trunk some issues with a missing gap, notably in TT1 blocks:
<img width="743" alt="Screenshot 2021-10-19 at 11 14 18" src="https://user-images.githubusercontent.com/1204802/137880326-42a6b3a6-7b9d-4cf1-aa0c-c20157bbc206.png">

It appears that the CSS variable fallback isn't working as expected for me, perhaps because the `calc` part of the definition returns a value even when the nested variable isn't there:

<img width="410" alt="Screenshot 2021-10-19 at 11 14 06" src="https://user-images.githubusercontent.com/1204802/137880440-3f924f5c-ea8a-4d87-9891-d2c55e353f73.png">

I happed upon [this excellent fallback strategy post](https://lea.verou.me/2021/10/custom-properties-with-defaults/) recently, in case it's useful for a followup.

jasmussen commented:
Oh, maybe I'm wrong. Maybe the problem here is that the gap is set to zero when the nav block has a background. 

This _could_ be related to the fact that in previous iterations, _menu items have padding_ — zero padding but a margin when there's no background, 0.5em padding and no margin when there is a background.

jasmussen commented:
As a followup, I'm looking into a fix. It appears there are two issues going on — both that the fallback value isn't used, but also that the padding appears to have been zeroed out.

jasmussen commented:
Created #35752 as a followup. I think I got everyone done, but I had to simplify the selector a bit. Happy to tweak based on your feedback.

ramonjd commented:
Thanks for the follow up on this one! 

> I happed upon this excellent fallback strategy post recently, in case it's useful for a followup.

Very timely, appreciate the link.
